File: Backtracking/Python/Split a String Into the Max Number of Unique Substrings.py
# Problem link: https://leetcode.com/problems/split-a-string-into-the-max-number-of-unique-substrings/description
# Time: O(n * 2 ^ n)
# Space: O(n)
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        seen = set()
        return self.dfs(s, 0, seen)
    
    def dfs(self, s, idx, seen):
        if idx == len(s):
            return len(seen)
        
        result = 0

        for i in range(idx, len(s)):
            prev = s[idx : i + 1]
            if prev not in seen:
                seen.add(prev)
                result = max(result, self.dfs(s, i + 1, seen))
                seen.remove(prev)
        
        return result

File: Backtracking/Java/Split a String Into the Max Number of Unique Substrings.java
// Problem link: https://leetcode.com/problems/split-a-string-into-the-max-number-of-unique-substrings/description
// Time: O(n * 2 ^ n)
// Space: O(n)
class Solution {
    public int maxUniqueSplit(String s) {
        Set<String> seen = new HashSet<>();
        return this.dfs(s, 0, seen);
    }

    private int dfs(String s, int idx, Set<String> seen) {
        if(idx == s.length()) {
            return seen.size();
        }

        int result = 0;

        for(int i = idx; i < s.length(); i++) {
            String strsub = s.substring(idx, i + 1);
            if(!seen.contains(strsub)) {
                seen.add(strsub);
                result = Math.max(result, this.dfs(s, i + 1, seen));
                seen.remove(strsub);
            }
        }

        return result;
    }
}
